src/Microsoft.Extensions.Primitives/StringSegment.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Extensions.Primitives
+{
+    public struct StringSegment : IEquatable<StringSegment>
+    {
+        private readonly string _buffer;
+        private readonly int _offset;
+        private readonly int _count;
+
+        public StringSegment(string buffer, int offset, int count)
```

Comment:
/cc @Tratcher @davidfowl - this constructor is used in HttpAbstractions with `offset = -1` to signify an invalid/null segment - this is a bit greasy to put on a public API.

Suggestion: Validate parameters here, require `buffer != null` and valid `offset` and `count` - use the default constructor for `invalid`.